syntax = "proto2";

package wsend;

option java_package = "com.wisight.wsend.proto";


message BotUpdate {
    //号码类型
    enum Type {
        PRIVATE = 1;    //私号
        PUBLIC = 2;     //公号 
    }

    //号码状态
    enum Status {
        INITIAL = 0;        
        TEMP_LOCKED = 2;
        OFFLINE_AUTO = 4;        
        BANNED = 6;
        KICKED = 7;
        OFFLINE_MANUAL = 8;
        ONLINE_AUTO  = 9;
        ONLINE_MANUAL = 5;
    }

    optional Type type = 1;
    optional int64 last_send_time = 2;
    optional Status status = 3;
    optional uint32 fail_reason_code = 4;  
    optional string ip = 5;                
    optional MsgLogItem msg_log = 6;
    optional int64 timestamp = 7;
    optional GroupsInfo groups_info = 8;    
    optional SyncInfo sync_info = 9;
    optional string task_id = 10;            
}

message SyncInfo {
    enum Mode {
        FULL = 1;    
        DELTA = 2;     
    }

    message User {
        optional string name = 1;   //通讯录名称
        optional string jid = 2;    //该联系人的jid
    }
    optional Mode mode = 1 ;      //full or delta
    repeated User users = 2;       
}


message GroupsInfo {
    enum GroupRole {
        MEMBER = 1;     
        OWNER = 2;      
        ADMIN = 3;      
    }
    message Group {
        message Participant {
            optional string jid = 1;
            optional GroupRole role = 2;

        }
        optional string id = 1;
        optional string subject = 2;
        optional string owner = 3;
        optional uint32 participant_count = 4;
        repeated Participant participants = 5;
    }

    optional string sender = 1;         
    optional int64 timestamp = 2;
    repeated Group groups = 3;
}


message MsgLogItem {
    enum Status {
        EXECUTED = 2;          
        SENT = 3;              
        RECEIVED = 4;          
        READ = 5;              
        RESPONSE = 6;          
        ERROR = 7;             
    }

    enum StatType {
        BS_SEND = 1;
        CS_SEND = 2; 
    }
    
    optional string msg_id = 1;
    optional string task_id = 2;
    optional string sender = 3;
    optional string target = 4;
    optional string content = 5;
    optional Status status = 6;
    optional int64 timestamp = 7;  
    optional StatType stat_type = 8;  
    optional string error_code = 9;     //只有在status=="ERROR"的时候才有这个值
}



message ClusterUpdate {    
    enum Status {
        OFFLINE = 1;       //离线事件
        ONLINE  = 2;       //上线事件
        ALIVE = 3;         //定期的保活事件
    }    
    optional Status status = 1;
    optional int64 timestamp = 2;            
}



message ContactUpdateItem {
    optional string target = 1;
    optional string key = 2;
    optional string value = 3;
}


message BotEvent {
    enum Event {
        LOGIN_SUCCESS = 1;
        LOGIN_FAIL    = 2;
        EXCEPTION     = 3;
        LOGOUT        = 4;
        QUIT          = 5;

        BOT_ERROR     = 6;

        CONTACT_UPDATE= 7;

        MSG_LOG      = 11;
    }
    optional string bot_id = 1;
    optional Event event = 2;    
    optional MsgLogItem msg_log= 3;
    optional int64 timestamp = 4;   
    
    optional string event_detail = 5;

    optional ContactUpdateItem contact_update = 6;
}


message Notification {

    enum Type {
        GROUP = 1;
    }

    message GroupNotification {

        enum Action {
            ADD = 1;
            REMOVE = 2;
        }

        optional string reason = 1;
        optional Action action = 2;
        repeated string jids = 3;
    }

    optional string id = 1;
    optional Type type =2;
    optional string sender =3;
    optional string target = 4;

    optional GroupNotification group_notification = 6;
}


message Message {
    optional string msg_id = 1;
    enum Type {
        TEXT = 1;
        IMAGE = 2;
        VIDEO = 3;
        AUDIO = 4;
        DOCUMENT = 5;
        BUTTONS = 6;
        LIST = 7;
        URL  = 8;
        BUTTONS_RESPONSE = 9;
        LIST_RESPONSE = 10;
        PRODUCT = 11;
        POLL = 12;
        POLL_RESPONSE = 13;
        STICKER = 14;
        REACTION = 15;
        AD = 16;
        OTHER = 99;

    }

    optional string sender = 2;
    optional string target = 3;    
    
    optional Type   type   = 4;
    optional int64  timestamp = 5;
    optional string task_id = 6;
    optional string bot_id = 7;
    optional string participant =8;     //群消息里面，sender是XXXX@g.us, 这个字段用来标识实际群消息里面的实际动作人的jid

    optional string notify=9;       //消息里面的notify名称
    
    oneof type_message{
        TextMessage text_message = 11;
        ImageMessage image_message = 12;
        VideoMessage video_message = 13;
        AudioMessage audio_message = 14;
        DocumentMessage document_message = 15;
        ButtonsMessage buttons_message = 16;
        ListMessage list_message = 17;
        UrlMessage  url_message = 18;
        ButtonsResponseMessage buttons_response_message = 19;
        ListResponseMessage list_response_message = 20;
        ProductMessage product_message = 21;
        PollCreateMessage poll_create_message = 22;
        PollUpdateMessage  poll_update_message = 23;
        StickerMessage sticker_message = 24;
        ReactionMessage reaction_message = 25;
        AdMessage ad_message = 26;
    }
}

message MessageKey {
    optional string remote_jid = 1 ;
    optional bool from_me = 2;
    optional string id = 3;
    optional string participant = 4 ;
}

message ReactionMessage {
  optional MessageKey key = 1;
  optional string text = 2;  
  optional string grouping_key = 3;
  optional int64 sender_timestamp_ms = 4;
}


message StickerMessage {
    optional string url = 1;
    optional bytes file_sha256 = 2;
    optional bytes file_enc_sha256 = 3;
    optional bytes media_key = 4;
    optional string mimetype = 5;
    optional uint32 height = 6;
    optional uint32 width = 7;
    optional string direct_path = 8;    
    optional uint64 file_length = 9;
    optional int64 media_key_timestamp = 10;
    optional uint32 first_frame_length = 11; 
    optional bytes first_frame_sidecar = 12;
    optional bool is_animated = 13;
    optional bytes png_thumbnail = 16 [deprecated=true];    
    optional int64 sticker_sent_ts = 18;
    optional bool is_avatar = 19;
    optional bool is_ai_sticker = 20;
    optional bool is_lottie = 21;
    optional string accessibility_label = 22;    
}

message PollCreateMessage {

    optional string name = 4;
    repeated string options = 5;
    
}

message PollUpdateMessage {
    optional string create_msg_id  = 1;
    repeated string selected = 6;
}

message ProductMessage {
    optional ImageMessage product_image = 1;
    optional string title = 2;
    optional string description = 3;
    optional string product_id = 4;
    optional string business_owner_jid = 5;

}

message ButtonsMessage {

    message Button {
        optional string button_id = 1;
        optional string display_text = 2;        
    }
    
    optional string text = 1;
    optional string description = 3;
    optional string footer = 4;
    repeated Button buttons = 9;
}

message ButtonsResponseMessage {
    optional string selected_id = 1;
    optional string selected_display_text = 2;
}

message ListMessage {

    message Row {
        optional string title = 1;
        optional string description = 2;
        optional string row_id = 3;
    }

    message Section {
        optional string title = 1;
        repeated Row rows = 2;
    }

    optional string description = 1;
    optional string title = 2;
    optional string button_text = 3;
    optional string footer = 4;

    repeated Section sections = 5;

}

message ListResponseMessage {
    optional string selected_row_id = 1;
    optional string title = 2;
    optional string description = 3;
}

message UrlMessage {
    optional string text = 1;
    optional string url  = 2;
    optional string title = 3;
    optional string description = 4;
}

message TextMessage {
    optional string text = 1;
}

message ImageMessage {
    optional string url = 1;
    optional string mimetype = 2;
    optional string caption = 3;
    optional bytes file_sha256 = 4;    
    optional uint64 file_length = 5;    
    optional uint32 height = 6;
    optional uint32 width = 7;
    optional bytes media_key = 8;
    optional bytes file_enc_sha256 = 9;
    optional string direct_path = 11;
    optional int64 media_key_timestamp = 12;
    optional bytes jpeg_thumbnail = 16;
}

message VideoMessage {

	enum Attribution {
		NONE = 0;
		GIPHY = 1;
		TENOR = 2;
	}    

    optional string url = 1;
    optional string mimetype = 2;
    optional bytes file_sha256 = 3;    
    optional uint64 file_length = 4;
    optional uint32 seconds = 5;
    optional bytes media_key = 6;
    optional string caption = 7;
    optional bool gif_playback = 8;
    optional uint32 height = 9;
    optional uint32 width = 10;
    optional bytes file_enc_sha256 = 11;
    optional string direct_path = 13;
    optional int64 media_key_timestamp = 14;

    optional bytes jpeg_thumbnail = 16;

 
}

message AudioMessage {

    optional string url = 1;
    optional string mimetype = 2;
    optional bytes file_sha256 = 3;
    optional uint64 file_length = 4;
    optional uint32 seconds = 5;
    optional bool ptt = 6;
    optional bytes media_key = 7;
    optional bytes file_enc_sha256 = 8;
    optional string direct_path = 9;
    optional int64 media_key_timestamp = 10;


}

message DocumentMessage {
    optional string url = 1;
    optional string mimetype = 2;
    optional string title = 3;
    optional bytes file_sha256 = 4;    
    optional uint64 file_length = 5;
    optional uint32 page_count = 6;
    optional bytes media_key = 7;
    optional string file_name = 8;
    optional bytes file_enc_sha256 = 9;
    optional string direct_path = 10;
    optional int64 media_key_timestamp = 11;
    optional bool contact_vcard = 12; 
    optional string thumbnail_direct_path = 13;
    optional bytes thumbnail_sha256 = 14;
    optional bytes thumbnail_enc_sha256 = 15;
    optional bytes jpeg_thumbnail = 16;
    optional uint32 thumbnail_height = 18;
    optional uint32 thumbnail_width = 19; 
    optional string caption = 20;    
}

message AdMessage {
    optional string text = 1;
    optional string title = 2;
    optional bytes  thumbnail = 3;
    optional string url = 4;
    optional bool   larger_thumbnail = 5;
}


message Command {
    optional string cmd_key = 1;
    optional string name = 3;
    repeated string params = 4;
    optional int64  timestamp = 5;
}



